### ISSUE TITLE
TLS 1.3: better handling for "close notify"

### ISSUE THREAD

**Issue by Anonymous-Coward** [envoyproxy/envoy#11619]
We run envoy as part of an istio gateway, in a pod in a kubernes cluster.

$ envoy --version
envoy  version: 9b4239dee83dd8894bfc579d412ccd894cff2597/1.13.1-dev/Clean/RELEASE/BoringSSL

TLS traffic terminates on envoy, everything downstream is plain http. The http port of envoy is also open, but not to the public.

When we send a request repeatedly through envoy using plain http, everything works fine.

When we send exactly the same request using https, also repeatedly, about 2%-3% of the responses are randomly truncated (i.e. not at the same length). In all other cases things work fine. There's no apparent periodicity or other regularity between when those truncations occur, or correlation with overall cluster or network load.

Since everything downstream is plain http, whether requests hit envoy over https or http should not make a difference, I believe, for how things work downstream. When we hit everything downstream with the same request directly, we never get an error.

We briefly enabled trace logging on envoy, but can't see anything suspicious in the logs.

What could cause this?

I can provide additional information on request, but don't know what to provide beforehand.

**Comment by htuch** [comment]
This seems mysterious, if you can share an anonymized PII-free stats dump then that might be a good start to look for anomalies.

**Comment by Anonymous-Coward** [comment]
[envoy-stats.txt](https://github.com/envoyproxy/envoy/files/4797279/envoy-stats.txt)

I hope the attached file contains what you wanted.

**Comment by Anonymous-Coward** [comment]
I had trace logging briefly enabled yesterday, while reproducing the error. It's difficult to catch the exact log fragment corresponding to the error, and a log fragment to include both processing resulting in both sane and truncated responses would be huge. Therefore, here's a fragment which (I suppose) corresponds to a sane response, maybe it's useful.
[trace-log-anonymized.txt](https://github.com/envoyproxy/envoy/files/4799796/trace-log-anonymized.txt)

On request, I can re-enable trace logging and attach the huge resulting log here.

**Comment by Anonymous-Coward** [comment]
Maybe also relevant: the k8s objects that make up our configuration - in the attachment.

How it all works: everything comes in over an istio gateway named very originally "external", and is routed in its entirety, via a virtual service, to a custom proxy which does custom stuff. The proxy then forwards it to another gateway, equally originally named "internal". From there onwards, via virtual services configured for each app individually, requests are routed to the apps responsible to process them.

I changed all app- and setup-specific names in the output. If something doesn't make sense, it's my mistake, and I can clarify things on request.

[app-specific-setup.txt](https://github.com/envoyproxy/envoy/files/4799918/app-specific-setup.txt)

**Comment by GregHanson** [comment]
```
':status', '477'
'content-security-policy', 'frame-ancestors 'none'; script-src 'self'; object-src 'none';'
'content-type', 'application/json'
'date', 'Wed, 17 Jun 2020 15:21:34 GMT'
'referrer-policy', 'no-referrer'
'server', 'istio-envoy'
'strict-transport-security', 'max-age=31536000; preload;'
'x-content-security-policy', 'frame-options 'deny';'
'x-content-type-options', 'nosniff'
'x-frame-options', 'DENY'
'x-xss-protection', '1; mode=block'
'x-envoy-upstream-service-time', '76'
```
@htuch the `477` code is returned by the destination app, but where are the `strict-transport-security` and `max-age=31536000` headers coming from?

**Comment by Anonymous-Coward** [comment]
Everything comes from envoy, it seems. The headers originally set on the response, by the application, are:

```
Content-Type: application/json
Transfer-Encoding: chunked
Date: Thu, 18 Jun 2020 15:32:54 GMT
```

The custom, non-standard http status code is indeed what the app returns.

**Comment by mattklein123** [comment]
Are you seeing timeouts? Timeouts can occur during a response when the headers and some body have already been forwarded.

**Comment by Anonymous-Coward** [comment]
No timeouts. I was suspecting this happening due to long response times from the app today, and checked. The response time is spread across several tens of ms - maybe 15 to 50. I couldn't notice a significant difference in spread between responses delivered whole and truncated responses.

**Comment by Anonymous-Coward** [comment]
At the suggestion of @GregHanson, attached the output of some istioctl commands, to give more insight into configuration.

Commands executed:

```
istioctl proxy-config routes <ingress_pod>.istio-system -o json > routes.json
istioctl proxy-config clusters <ingress_pod>.istio-system -o json > clusters.json
istioctl proxy-config listeners <ingress_pod>.istio-system -o json > listeners.json
```
Names of namespaces and pods and stuff were changed.

[istioctl-info.zip](https://github.com/envoyproxy/envoy/files/4805352/istioctl-info.zip)

**Comment by Anonymous-Coward** [comment]
I think I found the buggy behavior.

I captured packets and looked at them until I found a truncated response. Here's what happens:

- The connection gets established as http 2 over tls 1.3
- everything works as expected - the request is streamed, then the response is streamed back
- every once in a while, the client sends a close_notify message to the server (to envoy, that is), before it has fully received the response
- envoy, even if the tls version in use is 1.3, behaves as specified by tls v1.2 - immediately sends back a response close_notify, no longer sends any data, and also sends a tcp fin right after the response close_notify

The RFC for TLS v1.3 explicitly warns that the behavior specified in TLS v1.2 could cause truncation: "That previous requirement could cause truncation in the read side." - here, in the second paragaph: https://tools.ietf.org/html/rfc8446#section-6.1.

The TLS v1.2 behavior is specified here: https://tools.ietf.org/html/rfc5246#section-7.2.1

**Comment by linsun** [comment]
@htuch @mattklein123 @lizan @PiotrSikora thoughts?  Is this an envoy issue or if there is any configuration on envoy users can do to work around this?

Thanks much!

**Comment by lambdai** [comment]
I would suggest strip the other http filters if your analysis were not there.

An additional question: When truncate happens, what does the client see? A failed response?

**Comment by Anonymous-Coward** [comment]
We are not using any filters. I don't understand what you refer to.

The client just gets incomplete json. Failure happens on body parsing, not at the http connection level.

If you re-read my previous comment carefully, I think the problem will become obvious. It doesn't seem to have anything to do with filtering or other stuff. I'll try to explain again, in more detail:

- the TCP connection is established (the initial 3-way TCP handshake - SYN, SYN-ACK, ACK)
- the TLS version negotation is performed, and the client and envoy agree on TLS v1.3
- the request is sent by the client
- envoy relays the request
- envoy starts receiving the response
- envoy starts sending bits of the response back the the client
- while envoy has not yet finished sending all bits of the response back to the client, the client sends close_notify
- envoy, instead of finishing sending the remaining bits of the response, immediately responds with a close_notify of its own, and closes the TCP connection

Envoy's behavior, in the flow described above, corresponds to the TLS v1.2 spec, which says that upon receiving a close_notify envoy should stop sending the response. But the negotiated TLS version in use, at that point, is v1.3, and v1.3 has no such requirement. Instead, the v1.3 spec explicitly warns that in case of such behavior response truncation may occur.

I linked to the relevant sections in the two TLS specifications above.

**Comment by linsun** [comment]
@lambdai thank you for chiming in.  I noticed there is a few http filter related to Wasm that are newly introduced in Istio 1.5 (as default).  Would be good to test with these removed.

@Anonymous-Coward, if you run the following cmds you can see them.  It is installed as part of istio.

```
$ k get envoyfilter -n istio-system                                    
NAME                        AGE
metadata-exchange-1.4       14d
metadata-exchange-1.5       14d
metadata-exchange-1.6       14d
stats-filter-1.4            14d
stats-filter-1.5            14d
stats-filter-1.6            14d
tcp-metadata-exchange-1.5   14d
tcp-metadata-exchange-1.6   14d
tcp-stats-filter-1.5        14d
tcp-stats-filter-1.6        14d
```

**Comment by Anonymous-Coward** [comment]
Will try this on Monday. But I still don't expect this to make a difference.

**Comment by lambdai** [comment]
@Anonymous-Coward I am not a tls expert but I agree with you on the issue at TLS, which is an early shutdown though envoy is allowed to transmit the entire response.

My question is about a following up:
A http response has its own integrity check. In Http1 it's either response size nor chuncked encoding. In http2 it's stream flag.
These http mechanisms could help the http client to determine the response is a completed response, or a few bytes are supposed to follow on the wire if the TCP/TLS connection were not shutdown.

Are you saying the http client which talk to ingress-envoy doesn't report any error? My previous investigation was based on this assumption so I focused on the stream at ingress-envoy to its backend.

**Comment by Anonymous-Coward** [comment]
@lambdai I'm neither a TS nor a HTTP expert. From what I knew and what I looked up, except for the Content-length header, there's no way for the client to know if it has received the entire response, as long as chunks are whole - i.e. missing chunks at the end cannot be detected. The stream flag in http2 only allows avoiding mixing of streams, but again, not detecting if the end of the stream is missing. If I'm wrong, please correct me - as I said, I'm no expert.

What we provide is an API. We don't know the size of the response in advance - we start streaming it as soon as there are parts of it available. Buffering until we have the whole response, then starting to stream would hugely impact performance. Therefore we can't provide a content-length header.

**Comment by mattklein123** [comment]
> The TLS v1.2 spec states explicitly that after close_notify no more data should be sent, which implies that any stream, regardless of state, needs to be closed. With TLS v1.3 there's no such requirement.

I'm not familiar with the difference between TLS 1.2 and TLS 1.3 in this regard, but if there is some expected behavior difference between a TLS level "close notoify", we are almost certainly not handling that difference. We would need different types of events to be raised from the BoringSSL/TLS layer than what we have today I would imagine.

cc @davidben @PiotrSikora

**Comment by Anonymous-Coward** [comment]
My C/C++ is pretty rusty, and I don't have a properly set up C/C++ IDE, to easily browser code but I looked at the code anyway - both BoringSSL and envoy.

My very subjective and potentially wrong impression is that BoringSSL does not need changes. I'll explain.

In connection_impl.cc:

```
void ConnectionImpl::closeSocket(ConnectionEvent close_type) {
  ...
  transport_socket_->closeSocket(close_type);
   ...
}
```

In ssl_socket.cc:

```
void SslSocket::closeSocket(Network::ConnectionEvent) {
  ...
  if (state_ == SocketState::HandshakeInProgress || state_ == SocketState::HandshakeComplete) {
    shutdownSsl();
  }
}
```
`shutdownSsl();` should not be called if there are still open http 2 streams on the connection. BoringSSL/the transport layer has no awareness of https 2 streams on top of a connection, so there's no way it could signal anything.

**Comment by mattklein123** [comment]
> Once more: I might be completely off. If I am, please tell me and I'll delete my comment.

I think what you are saying is that you expecting half-close behavior to work correctly for HTTP. If so, yes, this does not work, and is tracked by https://github.com/envoyproxy/envoy/issues/2066.

**Comment by Anonymous-Coward** [comment]
@mattklein123  I read that other ticket. It only says about half-close at the application level with http 1.1. This here yields similar (incorrect) behavior, but when using http 2, and not as a result of application level signaling alone, but as the result of how transport level and application level protocols interact. Do you think an addition there might be helpful? I fear that a fix for either alone won't fix the other. (But I do think a solution is possible that solves both.)

**Comment by mattklein123** [comment]
AFAICT the fix is the same, we want to correctly deal with client half-closes and not immediately terminate the connection.

**Comment by Anonymous-Coward** [comment]
Depends on how it's done, IMO.

An event triggering a half-close might be received both at the application level and at the transport level protocols. More precisely, in the case of this bug, clients might trigger the connection half close both by sending a TLS (transport level) close_notify and by sending a http 2 (application level) GOAWAY. As long as both are treated similarly by envoy, and transition the connection into a half-closed state - and neither triggering an immediate connection close - that would solve both bugs. That's what I had in mind too when I talked about a solution that solves both.

I was fearing a solution where the ssl socket object and the connection object maintain/manage separate connection state. That would fix the http half-close bug, but leave this bug open. I'm not that deep into the code, yet, to know if that state is currently maintained separately or not.

**Comment by davidben** [comment]
BoringSSL doesn't impose any cross-read/write close_notify behavior and already behaves as in TLS 1.3. That is, when we receive close_notify, we continue to allow `SSL_write` and do not automatically trigger a close_notify of our own. The calling application may end up doing this, but that's on the calling application.

I don't think this interpretation is universal, but I see the TLS 1.3 change as fixing close_notify to align with reality and TCP semantics. TCP has a half-close state and some legacy HTTP/1.1 connection framings use it.

That said, it is surprising to me that you're observing a half-close in HTTP/2 clients. Part of the point of HTTP/2 was to improve connection reuse. If the client is doing connection reuse, it wouldn't close the connection yet.

**Comment by Anonymous-Coward** [comment]
It took a long while to reconfigure a cluster to allow for tcp dumps on all relevant pods, but after I was able to get tcp dumps I found envoy not to be the culprit. Some intermediate node behind already sends out the response truncated, so envoy has no way of sending out an un-truncated response.

I also wrote two small applications, client and server, where the client always sends out close_notify before receiving the full response from the server. With envoy running between the two, I was not able to reproduce the bug once.

The odd thing is that the bug on the intermediate node only happens when envoy is hit via https from outside of the k8s cluster.  If the client hitting envoy runs inside the envoy pod, or if envoy is hit via plain http, the bug never happens.  This is what made me suspect envoy, before getting tcp dumps - I had no reason to suspect envoy not getting the full response from the nodes behind it, ever.

This being said, I believe this bug report can be closed. I just don't want to do it myself, since I'm not a contributor to the project, so I don't know what the protocol is.
